element-view/src/androidMain/kotlin/ElementView.kt

```diff
@@ -0,0 +1,176 @@
+        }
+
+    override fun onAttachedToWindow() {
+        super.onAttachedToWindow()
+        adapter?.onAttached(this)
+    }
+
+    override fun onDetachedFromWindow() {
+        super.onDetachedFromWindow()
+        adapter?.onDetached()
+    }
+
+    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
+        adapter?.onSizeChanged(w, h)
+    }
+
+    override fun onDraw(canvas: Canvas) {
+        val bitmap = adapter?.bitmap
+        if (bitmap != null) {
+            canvas.drawBitmap(bitmap, 0f, 0f, blitPaint)
+        }
+    }
+
+    public class Adapter<T>(
+        private val dataSource: Flow<T>,
+        private val updater: Updater<T>,
+    ) {
+
+        /** Android's Size was added in API 21, so do this to avoid bumping min version. */
+        public data class State(val view: ElementView?, val root: RootElement, val width: Int, val height: Int)
+
+        public fun interface Updater<T> {
+            public fun update(root: RootElement, width: Float, height: Float, data: T)
+        }
+
+        private val state: MutableStateFlow<State> = MutableStateFlow(EMPTY_STATE)
+
+        /** Coroutine scope rendering occurs on. This scope is live between calls to [onAttached] and [onDetached]. */
+        private var renderScope: CoroutineScope? = null
+            set(value) {
+                if (field !== value) field?.cancel()
+                field = value
+            }
+
+        /** The most recent bitmap to finish rendering, if any. */
+        internal var bitmap: Bitmap? = null
+            private set
+
+        /**
+         * The view changed size, so we should re-render the bitmap. When this happens, the RootElement
+         * will be reset, so the render will have a clean slate.
+         */
+        internal fun onSizeChanged(width: Int, height: Int) {
+            state.value = state.value.copy(root = RootElement(), width = width, height = height)
+        }
+
+        /**
+         * Attach this to a view (which is itself attached to a window).
+         * This will immediately render the current state if possible.
+         */
+        internal fun onAttached(view: ElementView) {
+            state.value = State(view, RootElement(), view.width, view.height)
+            renderScope = object : CoroutineScope {
+                private val job = Job()
+                override val coroutineContext: CoroutineContext get() = job
+            }.also { launchRenderingIn(it) }
+        }
+
+        /** Detach this from a view (usually because the view was detached from a window). */
+        internal fun onDetached() {
+            renderScope = null
+            state.value = EMPTY_STATE
+        }
+
+        private fun launchRenderingIn(scope: CoroutineScope) = scope.launch {
+            // Unlimited buffer looks scary, but collection has way higher throughput than emission, so this doesn't grow large.
+            val buffer = MutableSharedFlow<Bitmap>(extraBufferCapacity = Int.MAX_VALUE)
```

Comment:
I think picking a buffer would be safer, then `tryEmit` on line 142 would simply drop frames if the producer was somehow faster than the consumer.

Otherwise I worry that library consumers might hit OOM failures, which can be hard to diagnose, I'd rather log frame drops when `tryEmit` returns `false`. 🤷 

_Now what that buffer should be, I don't know. Your guess is as good as mine._

Reply:
In my testing, I never got more than two items in the buffer. I'll cap it at 4 just to prevent any runaway that my test doesn't trigger, though.

Also going to use `BufferOverflow.SUSPEND` instead of dropping elements so that way the runaway renderer wouldn't keep eating CPU cycles that don't actually do anything.